### PR TITLE
Fix inverted array indexing and off-by-one errors in canvas.raster

### PR DESCRIPTION
**Pull request by philippjfr**
While testing the regridding implementation @jbednar identified a bug in the code, which were caused by array indices being inverted in certain cases. I tried for a long time to get the previous approach using affine transforms working but kept running into various off-by-one issues, which seemed to be caused by rounding issues in the affine transform code. I've now replaced this code with an approach that simply converts continuous coordinates to array indices by normalizing and then rescaling them. Not only do I understand this approach better but it now works consistently.

If you have a minute it would be good if you could review this @gbrener as I'm still trying to figure out what the affine transform approach gives us that's not handled by this simpler normalize and rescale approach.

### CODE REVIEW

File: datashader/tests/test_raster.py [hunk 1]
```diff
@@ -138,12 +138,12 @@ def test_raster_both_ascending_partial_range():
     ys = np.arange(5)
     arr = xs*ys[np.newaxis].T
     xarr = xr.DataArray(arr, coords={'X': xs, 'Y': ys}, dims=['Y', 'X'])
-    cvs = ds.Canvas(9, 4, x_range=(-.5, 8.5), y_range=(-.5, 3.5))
+    cvs = ds.Canvas(7, 3, x_range=(.5, 7.5), y_range=(.5, 3.5))
```
Comment (jbednar) [inline datashader/tests/test_raster.py]: Can you motivate this change to the tests?  Is it now testing something more than it was before?

Comment (philippjfr) [reply datashader/tests/test_raster.py]: Yes, it's now slicing different numbers of elements along each edge. Previously I was only slicing off the upper bound, which let some bugs slip past.

Comment (jbednar) [reply datashader/tests/test_raster.py]: Sounds good.

**Comment by philippjfr** [comment]
Thanks for the background @gbrener. We can revisit if we need to handle any other transformations supplied on the attributes.

**Pull request merged**